documentation/tutorial/intermittent_memory_spike.md

```diff
@@ -0,0 +1,179 @@
+
+What we would like this tool to do is allow the user to specify the pid of the target process as well as the memory consumption threshold (in MBs). It would then continuously monitor the process and create a dump if the threshold is breached:
+
+> ```bash
+> sudo ./triggerdump <pid> <memory threshold in MBs>
+> ```
+
+#### Some background before we start writing the tool...
+The dotnet core runtime contains a mechanism known as the EventPipe that carries events to interested consumers. There are several different events that flow through the EventPipe including diagnostics information such as counters. The EventPipe is exposed as a Unix domain socket on Linux machines and named pipes on Windows.  EventPipe is set to duplex mode which means that clients can both read and write to the pipe. A diagnostics application can register to consume these events from the EventPipe and create new diagnostics experiences. Rather than communicating directly with EventPipe there is a client library that can be used and implemented in Microsoft.Diagnostics.Tools.RuntimeClient.dll.
+
+Events that are written to the EventPipe can come from multiple sources (or providers) and as such, clients that receive events over EventPipe can filter those events based on specific providers.  
+
+#### Writing the tool...
+We have two requirements in order to implement a tool that will create a dump file based on memory consumption:
+
+* Read dotnet memory counter to know if the current memory consumptions goes above the specified threshold
+* Generate the actual core dump
+
+Let's start with the first requirement, reading dotnet counters. As explained earlier, we can use the EventPipe mechanism to read counters from the runtime. In this case, the provider that writes counter events is System.Runtime. Below is the code that 
+sets up the System.Runtime provider for use in our tool (for brevity, error checking has been excluded):
+
+```csharp
+CounterProvider provider=null;
+KnownData.TryGetProvider("System.Runtime", out provider);
+string prov = provider.ToProviderString(1); 
+```
+The above code attempts to get the System.Runtime provider which we will need in the next section of code that configures and starts the counter collection:
+
+```csharp
+Task monitorTask = new Task(() => 
+{
+    var configuration = new SessionConfiguration(circularBufferSizeMB: 1000, outputPath: "", providers: Trace.Extensions.ToProviders(prov.ToString()));
```

Comment:
It might be clearer to construct the array of providers directly from the constructor API rather than using Trace.Extensions.ToProviders() to parse a string.

Trace.Extensions isn't planned to be part of RuntimeClient - anyone taking a dependency on it is taking a dependency on implementation details of the dotnet-trace tool. dotnet-counters can get away with it because we build both tools in our repo and can coordinate their code but for anyone working outside the repo that won't work so well.